ops/testing.py

```diff
@@ -260,7 +260,20 @@ def update_relation_data(self, relation_id, app_or_unit, key_values):
             if rel_data is not None:
                 # If we have read and cached this data, make sure we invalidate it
                 rel_data._invalidate()
-        # TODO: we only need to trigger relation_changed if it is a remote app or unit
+
+        is_peer = self._charm.meta.relations[relation_name].role == 'peers'
```

Comment:
So in general, anything that *I* am doing doesn't trigger relation changed.
We could split it out into a different function 'update_my_relation_data' (for app or unit). I actually originally did most of that with init values:
def add_relation(self, app, *, initial_unit_data=None, initial_app_data=None, remote_app_data=None):

I got rid of that to simplify add_relation. But maybe we should bring that back, so you don't have the awkward first steps of:
```
harness = ...
rel_id = harness.add_relation(...)
harness.update_relation(rel_id, MYDATA)
harness.add_relation_unit(...)
```

Put another way. To update *your* data bucket, that is was 'relation_set' is for. So only allowing update_relation_data to be done for 'things that aren't you' might be more interesting, and instead putting the 'how do I set an initial value' into the 'add_relation' call.


Does that make sense?

Reply:
Yes, I agree that it feels awkward to be able to update your own unit's or app's data via other means than relation-set - it gives a false impression that this is doable in Juju in general.

`update_relation_data` does not give an indication about who is triggering the change but we always assume that it is a remote unit. If our unit is a leader, and our app relation data (peer or not) is updated via `update_relation_data`, an exception should be raised because one should use `relation.data[self.harness.charm.model.app]['mykey'] = 'newvalue'` to do those kinds of updates. Otherwise, it's a scenario where a leader unit have done it but its not our unit.

Reply:
Partially addressed in https://github.com/canonical/operator/pull/212/commits/516c13393c0c4d4222aa2d85f520d0e37ecb0860.

I still need to add initial_unit_data and initial_app_data but at least update_relation_data now raises `RuntimeError`s in case one tries to do things that Juju wouldn't normally do.

Reply:
Addressed in:

https://github.com/canonical/operator/pull/212/commits/72733e763536ee366bc4b41a5fd664411c3cabbe
https://github.com/canonical/operator/pull/212/commits/56f97408a11a835970080b0df1a8c069c7c2768e
https://github.com/canonical/operator/pull/212/commits/e466728e127a045ddca4628db98bd7c658ec6ddc